happi/audit.py

```diff
@@ -0,0 +1,424 @@
+"""
+This module defines the ``happi audit`` command line utility
+
+If the --file option is not provided, then it will take the path from
+Happi configuration file using the Client.find_config method,
+otherwise it will use the path specified after --file
+"""
+from abc import ABCMeta, abstractmethod
+from configparser import ConfigParser
+import logging
+import os
+import sys
+import subprocess
+import re
+from happi.client import Client
+from happi.loader import import_class, fill_template
+from happi.containers import registry
+from enum import Enum
+
+logger = logging.getLogger(__name__)
+
+
+class ReportCode(Enum):
+    """
+    Class to define report codes
+    """
+    SUCCESS = 0
+    INVALID = 1
+    MISSING = 2
+    EXTRAS = 3
+    NO_CODE = 9
+
+
+class Command(object, metaclass=ABCMeta):
+    """
+    Bluprint for the command class
+    This will be useful if all the commands are inheriting from it
+    """
+
+    def __init__(self, name, summary):
+        self.name = name
+        self.summary = summary
+
+    @abstractmethod
+    def add_args(self, parser):
+        """
+        Adds arguments to a parser
+
+        Parameters
+        -----------
+        parser: ArgumentParser
+            Parser to add arguments to
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def run(self, args):
+        """
+        Parses the arguments given to the command.
+        And hanles the logic for this command
+
+        Parameters
+        -----------
+            args: Namespace
+        """
+        raise NotImplementedError
+
+
+class Audit(Command):
+    """
+    Audits the database for valid entries.
+    """
+    def __init__(self):
+        self.name = 'audit'
+        self.help = "Inspect a database's entries"
+        self._all_devices = set()
+        self._all_items = []
+        # 9 - NO_CODE
+        self.report_code = ReportCode(9)
```

Comment:
```suggestion
        self.report_code = ReportCode.NO_CODE
```

When dealing with these enums - you should _never_ have to deal with the raw integer value, unless it's coming over the wire or something.

In this case, the values truly don't even matter: consider `enum.auto()` for the values and making the enum `__repr__` look something like `return self.name` to get a pretty repr.

Reply:
Ohh wow perfect, I did not know about the `auto()` option in Python - this is cool and exactly what I needed here, thanks a lot Ken - fixed it in 1f87a91, I hope is what you meant.

Reply:
Looks good overall, but isn't the repr you implemented exactly as the default one provided by `enum.Enum`?

https://github.com/pcdshub/happi/commit/1f87a9198276c31eb8e7fb68e79c277ff1c29678#diff-e967c494e6919641a643a09b9e94bddcR35-R36

Reply:
Probably, although when I do this: `print(repr(self.report_code.EXTRAS))` without the `__repr__()` I get the value as well, something like this: 
<ReportCode.EXTRAS: 4>

And  this: `return '<%s.%s>' % (self.__class__.__name__, self.name)` hides the value: 
<ReportCode.EXTRAS>

Regardless, I probably don't need to define it here right?